File: src/main/java/com/erikromson/datawallet/crypto/SealedBox.java
package com.erikromson.datawallet.crypto;

public final class SealedBox {

    static final int SEAL_OVERHEAD = 48;

    private SealedBox() {}

    public static byte[] seal(byte[] plaintext, byte[] recipientPublicKey) {
        byte[] ciphertext = new byte[plaintext.length + SEAL_OVERHEAD];
        int result = CryptoProvider.sodium().crypto_box_seal(
                ciphertext, plaintext, plaintext.length, recipientPublicKey
        );
        if (result != 0) {
            throw new CryptoException("crypto_box_seal failed");
        }
        return ciphertext;
    }

    public static byte[] open(byte[] ciphertext, byte[] publicKey, byte[] secretKey) {
        byte[] plaintext = new byte[ciphertext.length - SEAL_OVERHEAD];
        int result = CryptoProvider.sodium().crypto_box_seal_open(
                plaintext, ciphertext, ciphertext.length, publicKey, secretKey
        );
        if (result != 0) {
            throw new CryptoException("crypto_box_seal_open failed");
        }
        return plaintext;
    }
}

File: client/lib/src/crypto/sealed_box.dart
import 'dart:typed_data';

import 'package:sodium/sodium.dart';

import 'crypto_exception.dart';

/// Anonymous sealed-box encryption backed by libsodium (crypto_box_seal).
class SealedBox {
  SealedBox._();

  /// Encrypts [plaintext] for [publicKey] using an ephemeral sender key.
  ///
  /// The ciphertext is not reproducible (ephemeral key is random each call).
  static Uint8List seal(
      Sodium sodium, Uint8List plaintext, Uint8List publicKey) {
    return sodium.crypto.box.seal(message: plaintext, publicKey: publicKey);
  }

  /// Decrypts [ciphertext] using the recipient's [publicKey] and [secretKey].
  ///
  /// Throws [CryptoException] if authentication fails.
  static Uint8List open(Sodium sodium, Uint8List ciphertext,
      Uint8List publicKey, Uint8List secretKey) {
    final secKey = SecureKey.fromList(sodium, secretKey);
    try {
      return sodium.crypto.box.sealOpen(
          cipherText: ciphertext, publicKey: publicKey, secretKey: secKey);
    } catch (e) {
      throw CryptoException('SealedBox decryption failed', e);
    } finally {
      secKey.dispose();
    }
  }
}
